File: snippets/go/develop/durabletimers.go
package develop

import (
	"fmt"
	"time"

	restate "github.com/restatedev/sdk-go"
)

type DurableTimers struct{}

func (DurableTimers) Greet(ctx restate.Context, name string) (string, error) {
	// <start_here>
	if err := restate.Sleep(ctx, 10*time.Second); err != nil {
		return "", err
	}
	// <end_here>

	// <start_timer>
	sleepFuture := restate.After(ctx, 30*time.Second)
	callFuture := restate.Service[string](ctx, "MyService", "MyHandler").RequestFuture("hi")

	fut, err := restate.WaitFirst(ctx, sleepFuture, callFuture)
	if err != nil {
		return "", err
	}
	switch fut {
	case sleepFuture:
		if err := sleepFuture.Done(); err != nil {
			return "", err
		}
		return "sleep won", nil
	case callFuture:
		result, err := callFuture.Response()
		if err != nil {
			return "", err
		}
		return fmt.Sprintf("call won with result: %s", result), nil
	}
	// <end_timer>

	return "", nil
}

File: snippets/ts/src/develop/durable_timers.ts
import * as restate from "@restatedev/restate-sdk";
import { myService } from "./my_service";

const service = restate.service({
  name: "DurableTimers",
  handlers: {
    greet: async (ctx: restate.Context, name: string) => {
      // <start_sleep>
      await ctx.sleep({ seconds: 10 });
      // <end_sleep>

      // <start_timer>
      try {
        await ctx
          .serviceClient(myService)
          .myHandler("Hi")
          .orTimeout({ seconds: 5 });
      } catch (error) {
        if (error instanceof restate.TimeoutError) {
          console.error("Operation timed out:", error);
        } else {
          throw error; // Re-throw other errors
        }
      }
      // <end_timer>
    },
  },
});
